Controller/Index/Callback.php

```diff
@@ -0,0 +1,322 @@
+
+    /**
+     * @var ServiceHelper
+     */
+    protected $serviceHelper;
+
+    /**
+     * Callback constructor.
+     * @param Context $context
+     * @param JsonFactory $resultJsonFactory
+     * @param EventManager $eventManager
+     * @param ConfigHelper $configHelper
+     * @param Logger $logger
+     * @param ServiceHelper $serviceHelper
+     * @param RequestContentInterface $requestContent
+     * @param JsonFactory $jsonResultFactory
+     * @param OrderHelper $orderHelper
+     * @param OrderRepositoryInterface $magentoOrderRepo
+     * @param OrderRepository $orderRepository
+     * @SuppressWarnings(PHPMD.ExcessiveParameterList)
+     */
+    public function __construct(
+        Context $context,
+        JsonFactory $resultJsonFactory,
+        EventManager $eventManager,
+        ConfigHelper $configHelper,
+        Logger $logger,
+        ServiceHelper $serviceHelper,
+        RequestContentInterface $requestContent,
+        JsonFactory $jsonResultFactory,
+        OrderHelper $orderHelper,
+        OrderRepositoryInterface $magentoOrderRepo,
+        OrderRepository $orderRepository
+    ) {
+        parent::__construct($context, $resultJsonFactory, $eventManager, $configHelper, $logger);
+
+        $this->jsonResultFactory = $jsonResultFactory;
+        $this->orderHelper = $orderHelper;
+        $this->requestContent = $requestContent;
+        $this->magentoOrderRepo = $magentoOrderRepo;
+        $this->orderRepository = $orderRepository;
+        $this->serviceHelper = $serviceHelper;
+
+        $this->setEventName('callback');
+        $this->setEventMethod([$this, 'updatePaymentData']);
+    }
+
+    /**
+     * @return array|bool|ResponseInterface|ResultInterface|string
+     * @throws Exception
+     *
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     */
+    public function updatePaymentData()
+    {
+        $this->logger->debug(
+            sprintf(
+                "Callback controller is called with request: \n %s",
+                $this->requestContent->getContent()
+            )
+        );
+
+        $requestData = json_decode($this->requestContent->getContent(), true);
+
+        if (json_last_error() !== JSON_ERROR_NONE) {
+            return $this->createResult(
+                'error',
+                'Failed to JSON decode callback request data',
+                json_last_error() . "\nRequest Data:\n" . $this->requestContent->getContent()
+            );
+        }
+
+        $paymentData = null;
+        $transactionData = null;
+
+        foreach ($requestData as $requestKey => $requestValue) {
+            switch ($requestKey) {
+                case 'payment':
+                    $paymentData = $this->serviceHelper->getPaymentData($requestValue['id']);
+                    break;
+                case 'transaction':
+                    // Replaces specific transactions with generic 'transactions'
+                    $transactionUri = preg_replace(
```

Comment:
@asbjornu I totally understand your point. But we need to know the instrument type & transaction type before we make an API call to get the data. And since we're only getting the IDs from the callback, we decided to parse the IDs to extract the data.

About future updates of the SDK, I think if we version both the SDK & API, we can avoid future breaks. I suggest that as soon as you've latest updates in the API, flag it as a new version, keep the previous version of the API for a certain period & mark it as deprecated. And we can also update it the SDK with the support of backwards compatibility. In that way, we can prevent it from breaking in case of updates.

Reply:
You need to find another way to extract this information than parsing the URL. The URL could just as well be `https://api.payex.com/5b9ebb617990455f804ba53f1273ad23-3140f44a8b4247088fe054e944572f01`. The format of the URL doesn't matter and shouldn't matter to the client of the API. You have the `id` URL of the payment and you knew when the payment was created which type it was. The same applies to transactions.